### PR TITLE
Preprocessing cleanup

### PR DESCRIPTION
**Pull request by montaguegabe**
Based on the previous 2

### CODE REVIEW

File: cifar10-pytorch/tr_cg.py [hunk 1]
```diff
@@ -48,12 +55,11 @@ def train(args, model, device, train_loader_grad, train_loader_hess, optimizer,
 
         # Accumulate gradients of the loss
         loss = F.cross_entropy(output, target)
+        loss /= len(train_loader_grad)
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_cg.py]: Is this just to be consistent with Matlab? It doesn't make sense to me.

File: cifar10-pytorch/tr_steihaug.py [hunk 2]
```diff
@@ -29,10 +29,12 @@ def solve_tr_subprob(hess_fn, grad, delta, error_tol, max_iters, x0):
     while test_err > test_err_term and iters < max_iters and torch.norm(
             x) <= hat_delta:
 
+        iters += 1
+
         logger = logging.getLogger('trcg')
         logger.info('Attempt to solve Steihaug.')
 
-        if iters == 0:
+        if iters == 1:
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_steihaug.py]: Should this be synced with the new master?

Comment (montaguegabe) [reply cifar10-pytorch/tr_steihaug.py]: Are you talking about this change in particular? I don't understand

File: cifar10-pytorch/tr_cg.py (hunk 1 above)
Comment (montaguegabe) [reply cifar10-pytorch/tr_cg.py]: This divides by 10 since each gradient calculated in each of the 10 minibatches provides just 1/10th of the full gradient to be accumulated

File: cifar10-pytorch/tr_steihaug.py (hunk 2 above)
Comment (ByzanTine) [reply cifar10-pytorch/tr_steihaug.py]: One of your other PR did change this right? Do you need to merge with that PR (which I merged a moment ago into the master)?

File: cifar10-pytorch/tr_cg.py (hunk 1 above)
Comment (ByzanTine) [reply cifar10-pytorch/tr_cg.py]: Oh Ic, thanks for the explanation.

File: cifar10-pytorch/tr_steihaug.py [hunk 3]
```diff
@@ -7,14 +7,14 @@
 def solve_tr_subprob(hess_fn, grad, delta, error_tol, max_iters, x0):
 
     # Initialize guess
-    x = x0
+    x = x0.clone()
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_steihaug.py]: Ic, this x is modified and could affect value outside the function...

File: cifar10-pytorch/tr_steihaug.py (hunk 2 above)
Comment (montaguegabe) [reply cifar10-pytorch/tr_steihaug.py]: Ohh yeah will do that tomorrow. This derives from the one you merged and was meant to be merged into that one first I think.

**Comment by ByzanTine** [comment]
No, I understand the x = x0.clone. This is reasonable. Bacause this is a caller side reference. 

I don't understand why p,z,r are are cloned.

**Comment by montaguegabe** [comment]
It is just what is happening in the MATLAB code. Perhaps there is a way to remove one of the clone statements and have the behavior remain unaffected and we gain a minimal performance increase, but I was finding it very hard to test that and I was worried I would introduce another bug in the process

**Comment by montaguegabe** [comment]
So I just matched it exactly to MATLAB behavior

**Comment by ByzanTine** [comment]
I am unconvinced why variable inside a function need to be cloned. I can approve the x=x0 change, but  I think the others are panic programming. 

As the depute like this happens, normally I will separate the PR approved part. And leave the rest for discussion. 

And I do think x=x0.clone is crucial and should be checked in ASAP, but others need further discussion and caution.

**Comment by montaguegabe** [comment]
Sorry let me give a more clear example:

You can see the bottom of the steihaug method we have
```
rho_old = rho.clone()
...
rho = torch.dot(z, r)
```

Now imagine there is no`.clone()` call. The last line then would modify both the variables `rho` and `rho_old`. They point to the same memory and would remain totally equal after the operation. Now imagine the while loop runs a second time and we reach the line:

`beta = rho / rho_old`

Since they are still the same this is equivalent to writing:

`beta = 1`

However this is clearly not the desired behavior. We expect `rho` and `rho_old` to be different. Thus we need the clone statements, regardless of whether they are in or out of the function.

If we wish to avoid weird mutation side effects, cloning should be the default whenever we have assignment between tensors, just to be on the safe side. It is safe because it is more along the lines of functional programming, in which you try to reduce mutation as much as possible. Just another example, the line `r.add_(-alpha * w)` before was mutating the variable `p`, which does not appear in the expression.

File: cifar10-pytorch/tr_steihaug.py [hunk 4]
```diff
@@ -74,10 +76,9 @@ def solve_tr_subprob(hess_fn, grad, delta, error_tol, max_iters, x0):
         if torch.norm(x) >= hat_delta:
             break
 
-        rho_old = rho
-        z = r
+        rho_old = rho.clone()
```
Comment (ByzanTine) [inline cifar10-pytorch/tr_steihaug.py]: I think we can call copy_ for this line and the following. Especially the following, each clone will allocate a new piece of memory.

**Comment by ByzanTine** [comment]
Ok, I think now I understand this in and out, happy to merge!

**Comment by ByzanTine** [comment]
So most of the time my criteria are harsh if I don't understand the code.
If I understand the change inside and out, I am more lenient. 

If you can squash them, it would be better. Still, it's merge-able.
